packages/connect-form/src/hooks/connection-options-handler.ts

```diff
@@ -0,0 +1,175 @@
+import { ConnectionOptions } from "mongodb-data-service";
+import { MARKABLE_FORM_FIELD_NAMES } from "../constants/markable-form-fields";
+import { SSHFormErrors } from "../utils/connect-form-errors";
+import { ConnectFormState } from "./use-connect-form";
+import { checkForInvalidCharacterInHost } from "../utils/check-for-invalid-character-in-host";
+
+export type SSHConnectionOptions = NonNullable<ConnectionOptions['sshTunnel']>;
+
+export type SSHType = 'none' | 'password' | 'identity' | 'socks';
+
+export interface UpdateConnectionOptions {
+  type: 'update-connection-options';
+  currentTab: SSHType;
+  key: keyof SSHConnectionOptions;
+  value: string | number;
+}
+
+export function handleUpdateConnectionOptions(action: UpdateConnectionOptions, {
+  errors: initialErrors,
+  connectionStringUrl,
+  connectionOptions,
+  connectionStringInvalidError,
+  warnings,
+}: ConnectFormState): ConnectFormState {
+  const { key, value, currentTab } = action;
+
+  if (currentTab === 'none') {
+    return {
+      connectionStringUrl,
+      connectionOptions: {
+        connectionString: connectionStringUrl.toString(),
+        sshTunnel: undefined,
+      },
+      errors: initialErrors,
+      warnings,
+      connectionStringInvalidError,
+    };
+  }
+
+  const errors = validateSshOptions(
+    currentTab,
+    key,
+    value,
+    connectionStringUrl.isSRV,
+    connectionOptions.sshTunnel
+  );
+
+  if (!connectionOptions.sshTunnel) {
+    connectionOptions.sshTunnel = {} as SSHConnectionOptions;
+  }
+
+  connectionOptions.sshTunnel[key] = value;
```

Comment:
I think we probably want to guarantee what's passed to this function isn't edited, rather we return a new object with the updated values. How do you feel about doing this below in the return? Maybe something like
```ts
const response: ConnectFormState = {
    connectionStringUrl,
    connectionOptions: {
      ...connectionOptions,
      sshTunnel: {
        ...connectionOptions.sshTunnel,
        [`${key}`]: value
      },
      connectionString: connectionStringUrl.toString(),
    },
    ...
 ```